cozy/codegen/rpython.py

```diff
@@ -0,0 +1,422 @@
+"""
+Basic generation of RPython.
+
+The generated code will depend on rlib.
+"""
+
+from contextlib import contextmanager
+from collections import defaultdict
+
+from cozy.codegen.misc import CodeGenerator
+from cozy.syntax_tools import subst
+from cozy.target_syntax import (TBool, TInt,
+        EBool, EEmptyList, ESingleton)
+
+class RPythonPrinter(CodeGenerator):
+
+    def __init__(self, *args, **kwargs):
+        super(RPythonPrinter, self).__init__(*args, **kwargs)
+
+        self.needs = defaultdict(bool)
+        self.instance_names = set()
+
+    @contextmanager
+    def python_block(self, *args):
+        """
+        Write a block delimited in the Python style by a trailing colon.
+        """
+
+        self.begin_statement()
+        self.write(*args, ":")
+        self.end_statement()
+        with self.indented():
+            yield
+
+    @contextmanager
+    def parens(self):
+        self.write("(")
+        yield
+        self.write(")")
+
+    @contextmanager
+    def stmt(self):
+        self.begin_statement()
+        yield
+        self.end_statement()
+
+    def visit_ENum(self, node):
+        self.write(str(node.val))
+
+    def visit_EBool(self, node):
+        self.write("True" if node.val else "False")
+
+    def visit_EEmptyList(self, _):
+        self.write("[]")
+
+    def visit_ESingleton(self, node):
+        self.write("[")
+        self.visit(node.e)
+        self.write("]")
+
+    def visit_EVar(self, node):
+        name = node.id
+        if name in self.instance_names:
+            self.write("self.", name)
+        else:
+            self.write(name)
+
+    def emit_unary(self, op, e):
+        self.needs[op] = True
+        self.write("_", op, "(")
+        self.visit(e)
+        self.write(")")
+
+    def visit_EUnaryOp(self, node):
+        op = node.op
+        if op in ("distinct", "reversed", "sum"):
```

Comment:
Ah, one more thing: there are proper constants for these names in `syntax.py`. It might be slightly preferable to use those.

Reply:
Definitely.

For some of these, would it be possible to get more broken-out nodes? I don't know what that does to the Cozy engine, but for some ops, like `ESingleton`, the broken-out unary op here seems to work well. On the other hand, it could be the case that keeping them together makes sense for other backends.

One more thing: Thank you for the review! I really appreciate it. For now, we're looking at replacing some of [Typhon](/monte-language/typhon)'s mutable structures with Cozy-generated ones, and hopefully then branching out into generating Monte code. It's unfortunate that we probably won't be able to use Cozy for our core structures, but maybe I'll be saved from writing yet another HTTP parser? The future is mysterious indeed.

Peace! ~ C.